File: Level2/배달.py
def solution(N, road, K):
    vilage = [[1e8] * N for _ in range(N)]
    answer = 1

    for idx in road:
        if vilage[idx[0] - 1][idx[1] - 1] == 0 and vilage[idx[1] - 1][idx[0] - 1] == 0:
            vilage[idx[0] - 1][idx[1] - 1] = idx[2]
            vilage[idx[1] - 1][idx[0] - 1] = idx[2]
        elif vilage[idx[0] - 1][idx[1] - 1] > 0 and vilage[idx[1] - 1][idx[0] - 1] > 0:
            vilage[idx[0] - 1][idx[1] - 1] = min(vilage[idx[0] - 1][idx[1] - 1], idx[2])
            vilage[idx[1] - 1][idx[0] - 1] = min(vilage[idx[0] - 1][idx[1] - 1], idx[2])

    for path in range(N):
        for start in range(N):
            for end in range(N):
                vilage[start][end] = min(vilage[start][end], vilage[start][path] + vilage[path][end])

    for idx in range(1, len(vilage[0])):
        if 0 < vilage[0][idx] <= K:
            answer += 1

    return answer

File: Level2/배달.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public class Node implements Comparable<Node> {
        int num;
        int cost;
        Node(int num, int cost) {
            this.num = num;
            this.cost = cost;
        }
        @Override 
        public int compareTo(Node e) {
            return this.cost - e.cost;
        }
    }
    
    public int solution(int N, int[][] road, int K) {
        ArrayList<Node>[] arr = new ArrayList[N+1];
        for(int i=1; i<=N; i++)
            arr[i] = new ArrayList<>();
        
        int[] costs = new int[N+1];
        Arrays.fill(costs, Integer.MAX_VALUE);
        
        for(int[] r : road) {
            arr[r[0]].add(new Node(r[1], r[2]));
            arr[r[1]].add(new Node(r[0], r[2]));
        }
        
        PriorityQueue<Node> queue = new PriorityQueue<>();
        queue.add(new Node(1, 0));
        
        while(!queue.isEmpty()) {
            Node node = queue.poll();
            int dest = node.num;
            for(Node nd : arr[dest]) {
                int nwDest = nd.num;
                if(costs[nwDest] > node.cost + nd.cost) {
                    costs[nwDest] = node.cost + nd.cost;
                    queue.add(new Node(nwDest, costs[nwDest]));
                }
            }
        }
        
        int answer = 1;
        for(int i=2; i<=N; i++) 
            if(costs[i] <= K) 
                answer++;
        
        return answer;
    }
}
